hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/physical/JoinHashPhysicalRel.java

```diff
@@ -58,9 +58,29 @@ public Join copy(
         return new JoinHashPhysicalRel(getCluster(), traitSet, left, right, conditionExpr, joinType);
     }
 
+    /**
+     * Cost calculation of Hash Join relation. It does not rely on children cost.
+     * <p>
+     * Hash Join algorithm is a more advanced join algorithm, which builds a hash table for the left
+     * row set, and then compare each row from the right side. Cost estimation is the following: <ol>
+     * <li> Processed row count (PR) is L + R because we traverse both sides once per join.
+     * <li> CPU is L * (hash table build cost) + R * (row comparison cost) + (PR * (row projection cost))</ol>
+     * <p>
+     * A perfect estimation must also include memory (occupied by the hash table) and IO costs.
+     */
     @Override
     @Nullable
     public RelOptCost computeSelfCost(RelOptPlanner planner, RelMetadataQuery mq) {
-        return super.computeSelfCost(planner, mq).multiplyBy(COST_FACTOR);
+        double leftRowCount = mq.getRowCount(getLeft());
+        double rightRowCount = mq.getRowCount(getRight());
+
+        double projectionCost = getRight().getRowType().getFieldCount();
+
+        double processedRowsCount = leftRowCount + rightRowCount;
+        double cpu = leftRowCount * Cost.HASH_JOIN_MULTIPLIER
```

Comment:
in case of outer join one of the sides is broadcast. This requires more processing. Shouldn't we include that in cost calculation?

Reply:
> Shouldn't we include that in cost calculation

Yes, makes sense.

Reply:
https://github.com/hazelcast/hazelcast/pull/25249/commits/f65293f63779688a039ff62fe67a27ec787b7447

Reply:
This is not what I meant: in case of outer non-equi join, outer side is broadcast. So we do `rightCount*memberCount` lookups which directly translates to CPU cost of lookups. However, they are also different than in case of equi-join. Eg. for query like `select * from m1 left join m2 on m1.__key<>m2.__key` hashmap in `SqlHashJoinP` degenerates to single key, and we iterate over all rows (this is a de-facto nested loop with spooling left side in memory)

Additionally, those rows have to be sent over network, but AFAIR we do not add network cost to CPU cost. We have separate `network` cost, which has it's own mutliplier but is generally `0`.

Reply:
> hashmap in SqlHashJoinP degenerates to single key

If I see correctly in case of non-equi hash join the hashmap is always degenerate. So we:
1. build `memberCount` single-unique-key hash maps (multimaps to be precise) each containing `rightCount` items
2. then we do `leftCount` lookups (in single-key hash map) and iterate over all found rows (ie. `rightCount` rows). This gives `leftCount` lookups (are they constant time with regard to `rightCount`?) and `leftCount*rightCount` comparisons for iterating over elements found in multimap (ie. all right side rows).
3. then projection of matching rows

Reply:
equi-join case is somewhere on the continuum between "hashmap lookup gives always single right row" and "hashmap lookup gives all right rows" depending on selectivity of equijoin columns (which we cannot estimate because we do not have column histograms).
also in equi-join hash table is partitioned, but might be skewed or even degenerate for the same reason

Reply:
>  lookups which directly translates to CPU cost of lookups. However, they are also different than in case of equi-join. Eg. for query like select * from m1 left join m2 on m1.__key<>m2.__key hashmap in SqlHashJoinP degenerates to single key.

Hmmmmm... Maybe don't even try to probe LEFT JOINs to be a hash joins...